src/NuGet.Core/NuGet.Packaging/Signing/Archive/EndOfCentralDirectoryRecord.cs

```diff
@@ -47,5 +46,65 @@ internal static EndOfCentralDirectoryRecord Read(BinaryReader reader)
 
             return header;
         }
+
+        private static void SeekToEndOfCentralDirectoryRecord(BinaryReader reader)
+        {
+            var byteSignature = BitConverter.GetBytes(Signature);
+            var length = reader.BaseStream.Length;
+
+            if (length < byteSignature.Length)
+            {
+                ThrowByteSignatureNotFoundException(byteSignature);
+            }
+
+            const int DefaultBufferSize = 4096;
+
+            var bufferSize = (int)Math.Min(length, DefaultBufferSize);
+            var matchingByteCount = 0;
+
+            // position is 0-based, while bufferSize is 1-based.
+            for (var position = length - bufferSize; position >= 0; position -= bufferSize)
+            {
+                reader.BaseStream.Position = position;
+
+                var buffer = reader.ReadBytes(bufferSize);
+
+                for (var i = buffer.Length - 1; i >= 0; --i)
+                {
+                    if (buffer[i] == byteSignature[byteSignature.Length - matchingByteCount - 1])
+                    {
+                        ++matchingByteCount;
+
+                        if (matchingByteCount == byteSignature.Length)
+                        {
+                            reader.BaseStream.Position = position + i;
+
+                            return;
+                        }
+                    }
+                    else
+                    {
+                        matchingByteCount = 0;
+                    }
+                }
+
+                bufferSize = (int)Math.Min(position, bufferSize);
```

Comment:
isn't you just need to check if position has reached to 0 or not? Will bufferSize ever be 0?

Reply:
That would work if the stream length were a multiple of `bufferSize`.  However, if it's not, you still need to process the remaining bytes.

Suppose you had a stream 5 bytes long and your buffer size is 3 bytes.  You go back from the end of the stream 3 bytes and read 3 bytes.  Then you need to read the remaining 2 bytes.  The next buffer size will match the remaining byte count:  2.